freedomofpress/securedrop
# Attempts to fix #2067 Pagination

## Status

Ready for review

## Description of Changes

Fixes #2067 

- Paginates the list of sources in journalist interface
- Reduces SD logo redrawing during pageload (better UX)
- Handles paginated sorting of starred items
- Doesn't lose your spot when you star or unstar a source

## Testing

- Increase num_sources in create-dev-data.py to at least 15 (if you want to hand-edit `journalist_app/main.py` to set `per_page=50` to something like 1) or several hundred (if you don't want to edit that line, and have the time/cpu)
- Login as a journalist
- Note the pagination buttons along the bottom
- Click a button or visit the ?page=2 page
- Note items are paginated as expected
- Visit the ?page=0 and ?page=999 pages
- Note that a 404 error is returned since there are not that many pages
- Visit Page 2 and Star a source. Note that the starred item disappears (moves to the top of the list on page 1) but your view remains on Page 2.
- Stay on Page 2 and use some of the mass-edit buttons at the top. Note that they do not keep your view on Page 2, which *may or may not be expected behavior.*
- Use some of the filtering, mass-select, download, star, or delete functions at the top. Note that many of them only apply to sources visible on the current page, which *may or may not be expected behavior.*

## Deployment

I intentionally stayed away from adding dependencies or configuration requirements.

## Checklist

### If you made changes to the server application code:


### If you made non-trivial code changes:
